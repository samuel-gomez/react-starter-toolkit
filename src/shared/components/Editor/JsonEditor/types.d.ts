// reproduce this type because the library doesn't export it on definition type file
// to be probably remove later

export type JsonEditorProps = {
  content?: Content;
  readOnly?: boolean;
  indentation?: number | string;
  tabSize?: number;
  mode?: Mode;
  mainMenuBar?: boolean;
  navigationBar?: boolean;
  statusBar?: boolean;
  escapeControlCharacters?: boolean;
  escapeUnicodeCharacters?: boolean;
  parser?: JSONParser;
  validator?: Validator | null;
  validationParser?: JSONParser;
  pathParser?: JSONPathParser;
  queryLanguages?: QueryLanguage[];
  queryLanguageId?: string;
  onChangeQueryLanguage?: OnChangeQueryLanguage;
  onChange?: OnChange;
  onRenderValue?: OnRenderValue;
  onClassName?: OnClassName;
  onRenderMenu?: OnRenderMenu;
  onChangeMode?: OnChangeMode;
  onError?: OnError;
  onFocus?: OnFocus;
  onBlur?: OnBlur;
  get?: () => Content;
  set?: (newContent: Content) => void;
  update?: (updatedContent: Content) => void;
  patch?: (operations: JSONPatchDocument) => JSONPatchResult;
  expand?: (callback?: (path: JSONPath) => boolean) => void;
  transform?: (options: TransformModalOptions) => void;
  validate?: () => ContentErrors;
  acceptAutoRepair?: () => Content;
  scrollTo?: (path: JSONPath) => void;
  findElement?: (path: JSONPath) => Element;
  focus?: () => void;
  refresh?: () => void;
  updateProps?: (props: JSONEditorPropsOptional) => void;
  destroy?: () => void;
};
