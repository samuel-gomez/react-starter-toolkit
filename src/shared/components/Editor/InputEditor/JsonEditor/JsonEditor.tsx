import { Button } from '@axa-fr/react-toolkit-all';
import Modal from '@axa-fr/react-toolkit-modal-default';
import ReactCodeEditor from '@uiw/react-textarea-code-editor';
import '@uiw/react-textarea-code-editor/dist.css';
import { ChangeEvent, useCallback, useState } from 'react';
import { ModalCommonHeader, ModalCommonBody, ModalCommonFooter, useToggleModal } from 'shared/components/ModalCommon';
import { TEvent, TonChange } from '../../Editor';
import './JsonEditor.scss';

type TuseJsonEditor = TEvent & {
  onChange: TonChange;
};

export const useJsonEditor = ({ value, id, name, onChange }: TuseJsonEditor) => {
  const [code, setCode] = useState(value);

  const onChangeJsonEditor = useCallback(
    (evn: ChangeEvent<HTMLTextAreaElement>) => {
      onChange({ value: JSON.parse(evn.target.value), name, id });
      setCode(JSON.parse(evn.target.value));
    },
    [id, name, onChange],
  );

  return { onChangeJsonEditor, code, setCode };
};

export type TReturnUseJsonEditor = ReturnType<typeof useJsonEditor>;

type TJsonEditor = TuseJsonEditor & {
  useJsonEditorFn?: typeof useJsonEditor;
  useToggleModalFn?: typeof useToggleModal;
  labelBtnOpenCodeEditor?: string;
};

const JsonEditor = ({
  value,
  onChange,
  name,
  id,
  labelBtnOpenCodeEditor = 'Edit params',
  useJsonEditorFn = useJsonEditor,
  useToggleModalFn = useToggleModal,
}: TJsonEditor) => {
  const { code, onChangeJsonEditor } = useJsonEditorFn({ value, onChange, name, id });
  const { onCancel, openModal, isOpen } = useToggleModalFn();

  return (
    <>
      <Button classModifier="editor" type="submit" onClick={openModal}>
        <span className="af-btn__text">{labelBtnOpenCodeEditor}</span>
      </Button>
      <Modal isOpen={isOpen} onOutsideTap={onCancel} className="af-modal af-modal--editorjson">
        <ModalCommonHeader onCancel={onCancel} title={`Saisir la value de ${name}`} />
        <ModalCommonBody>
          <ReactCodeEditor
            value={JSON.stringify(code, null, 2)}
            language="json"
            placeholder="Please enter JSON code."
            onChange={onChangeJsonEditor}
            padding={15}
            style={{
              fontSize: 12,
              fontFamily: 'ui-monospace,SFMono-Regular,SF Mono,Consolas,Liberation Mono,Menlo,monospace',
              minHeight: '200px',
            }}
          />
        </ModalCommonBody>
        <ModalCommonFooter cancelLabel="Fermer" onCancel={onCancel} />
      </Modal>
    </>
  );
};

export default JsonEditor;
