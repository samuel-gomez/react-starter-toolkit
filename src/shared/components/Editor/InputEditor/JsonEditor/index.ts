export { default, useJsonEditor } from './JsonEditor';
export type { TReturnUseJsonEditor } from './JsonEditor';
