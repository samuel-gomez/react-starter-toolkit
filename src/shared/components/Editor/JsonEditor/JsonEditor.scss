@import '@axa-fr/react-toolkit-core/dist/assets/scss/core.scss';

.af-modal--editor {
  width: 60vw;
  max-width: inherit;
}

@include media-breakpoint-down(sm) {
  .af-modal--editor {
    width: 100vw;
  }
}
