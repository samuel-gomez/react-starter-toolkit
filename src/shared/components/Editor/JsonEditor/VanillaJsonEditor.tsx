import { JSONEditor, type JSONEditorPropsOptional } from 'vanilla-jsoneditor';
import { useEffect, useRef } from 'react';
import type { JsonEditorProps } from './types';

export default function VanillaJSONEditor(props: JSONEditorPropsOptional) {
  const refContainer = useRef<HTMLDivElement>(null);
  const refEditor = useRef<JsonEditorProps | null>(null);

  useEffect(() => {
    // create editor
    refEditor.current = new JSONEditor({
      target: refContainer.current as Element,
      props: {
        mainMenuBar: false,
        navigationBar: false,
      },
    });

    return () => {
      // destroy editor
      if (refEditor.current) {
        refEditor?.current?.destroy?.();
        refEditor.current = null;
      }
    };
  }, []);

  // update props
  useEffect(() => {
    if (refEditor.current) {
      console.log('update props', props);
      refEditor?.current?.updateProps?.(props);
    }
  }, [props]);

  return <div className="vanilla-jsoneditor-react" ref={refContainer}></div>;
}
