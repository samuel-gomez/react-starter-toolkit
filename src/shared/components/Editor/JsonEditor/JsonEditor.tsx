import { Button } from '@axa-fr/react-toolkit-all';
import Modal from '@axa-fr/react-toolkit-modal-default';
import type { JSONContent, OnChange } from 'vanilla-jsoneditor';
import '@uiw/react-textarea-code-editor/dist.css';
import { useCallback, useState } from 'react';
import { ModalCommonHeader, ModalCommonBody, ModalCommonFooter, useToggleModal } from 'shared/components/ModalCommon';
import { TEvent, TonChange } from '../Editor';
import VanillaJsonEditor from './VanillaJsonEditor';
import './JsonEditor.scss';

type TuseJsonEditor = TEvent & {
  onChange: TonChange;
};

export const useJsonEditor = ({ value, id, name, onChange }: TuseJsonEditor) => {
  const [code, setCode] = useState({
    json: value,
  } as JSONContent);

  const onChangeJsonEditor = useCallback(
    (...params: Parameters<typeof setCode>) => {
      const newValue = params?.[0] as JSONContent;
      onChange({ value: newValue.json as unknown as string, name, id });
      return setCode(...params);
    },
    [id, name, onChange],
  );

  return { onChangeJsonEditor, code, setCode };
};

export type TReturnUseJsonEditor = ReturnType<typeof useJsonEditor>;

type TJsonEditor = TuseJsonEditor & {
  useJsonEditorFn?: typeof useJsonEditor;
  useToggleModalFn?: typeof useToggleModal;
  labelBtnOpenCodeEditor?: string;
};
const JsonEditor = ({
  value,
  onChange,
  name,
  id,
  labelBtnOpenCodeEditor = 'Edit params',
  useJsonEditorFn = useJsonEditor,
  useToggleModalFn = useToggleModal,
}: TJsonEditor) => {
  const { code, onChangeJsonEditor } = useJsonEditorFn({ value, onChange, name, id });

  const { onCancel, openModal, isOpen } = useToggleModalFn();
  return (
    <>
      <Button type="submit" onClick={openModal}>
        <span className="af-btn__text">{labelBtnOpenCodeEditor}</span>
      </Button>
      <Modal isOpen={isOpen} onOutsideTap={onCancel} className="af-modal af-modal--editor">
        <ModalCommonHeader onCancel={onCancel} title={`Saisir la value de ${name}`} />
        <ModalCommonBody>
          <VanillaJsonEditor content={code} onChange={onChangeJsonEditor as OnChange} />
        </ModalCommonBody>
        <ModalCommonFooter cancelLabel="Fermer" onCancel={onCancel} />
      </Modal>
    </>
  );
};

export default JsonEditor;
