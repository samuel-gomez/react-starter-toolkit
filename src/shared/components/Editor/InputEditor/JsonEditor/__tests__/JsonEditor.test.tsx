import { act, render, renderHook } from '@testing-library/react';
import { ChangeEvent } from 'react';
import JsonEditor, { useJsonEditor } from '..';

const onChange = jest.fn();
const defaultProps = {
  onChange,
  value: 'value',
  name: 'name',
  id: 'id',
};

describe('<JsonEditor/>', () => {
  it('Should render JsonEditor', () => {
    const { asFragment } = render(<JsonEditor {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('useJsonEditor', () => {
  const event = {
    preventDefault: jest.fn(),
    target: {
      value: `[
      {
        "label": "PrÃ©nom",
        "id": "firstname"
      }
    ]`,
    },
  };

  it('Should update code state when onChangeJsonEditor have been called with event that contain new value', () => {
    const { result } = renderHook(() => useJsonEditor({ ...defaultProps }));
    act(() => result.current.onChangeJsonEditor(event as unknown as ChangeEvent<HTMLTextAreaElement>));
    expect(result.current.code).toEqual(JSON.parse(event.target.value));

    expect(onChange).toHaveBeenCalledWith({ value: JSON.parse(event.target.value), name: 'name', id: 'id' });
  });
});
