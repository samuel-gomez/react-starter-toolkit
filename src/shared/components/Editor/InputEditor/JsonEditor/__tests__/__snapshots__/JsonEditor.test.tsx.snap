// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JsonEditor/> Should render JsonEditor 1`] = `
<DocumentFragment>
  <button
    class="btn af-btn af-btn--editor"
    type="submit"
  >
    <span
      class="af-btn__text"
    >
      Edit params
    </span>
  </button>
</DocumentFragment>
`;
